backend/src/hatchling/builders/wheel.py

```diff
@@ -282,6 +307,78 @@ def core_metadata_constructor(self) -> Callable[..., str]:
+                    VARIANT_INFO_DEFAULT_PRIO_KEY: {},
+                    VARIANT_INFO_PROVIDER_DATA_KEY: {},
+                    VARIANTS_JSON_VARIANT_DATA_KEY: {}
+                }
+
+                # ==================== VARIANT_INFO_DEFAULT_PRIO_KEY ==================== #
+
+                if (ns_prio := variant_config.default_priorities["namespace"]):
+                    data[VARIANT_INFO_DEFAULT_PRIO_KEY][VARIANT_INFO_NAMESPACE_KEY] = ns_prio
+
+                if (feat_prio := variant_config.default_priorities["feature"]):
+                    data[VARIANT_INFO_DEFAULT_PRIO_KEY][VARIANT_INFO_FEATURE_KEY] = feat_prio
+
+                if (prop_prio := variant_config.default_priorities["property"]):
+                    data[VARIANT_INFO_DEFAULT_PRIO_KEY][VARIANT_INFO_PROPERTY_KEY] = prop_prio
+
+                if not data[VARIANT_INFO_DEFAULT_PRIO_KEY]:
+                    # If no default priorities are set, remove the key
+                    del data[VARIANT_INFO_DEFAULT_PRIO_KEY]
+
+                # ==================== VARIANT_INFO_PROVIDER_DATA_KEY ==================== #
+
+                variant_providers = defaultdict(dict)
+                for ns, provider_cfg in variant_config.providers.items():
+                    variant_providers[ns][VARIANT_INFO_PROVIDER_REQUIRES_KEY] = provider_cfg.requires
+
+                    if provider_cfg.enable_if is not None:
+                        variant_providers[ns][VARIANT_INFO_PROVIDER_ENABLE_IF_KEY] = provider_cfg.enable_if
+
+                    if provider_cfg.optional:
+                        variant_providers[ns][VARIANT_INFO_PROVIDER_OPTIONAL_KEY] = True
+
+                    if provider_cfg.plugin_api is not None:
+                        variant_providers[ns][VARIANT_INFO_PROVIDER_PLUGIN_API_KEY] = provider_cfg.plugin_api
+
+                data[VARIANT_INFO_PROVIDER_DATA_KEY] = variant_providers
+
+                # ==================== VARIANTS_JSON_VARIANT_DATA_KEY ==================== #
+
+                variant_data = defaultdict(lambda: defaultdict(set))
+                for vprop_str in (variant_config.properties or []):
+                    match = VALIDATION_PROPERTY_REGEX.match(vprop_str)
+                    if not match:
+                        raise ValueError(
+                            f"Invalid variant property '{vprop_str}' in variant {variant_config.variant_hash}"
+                        )
+                    namespace = match.group('namespace')
+                    feature = match.group('feature')
+                    value = match.group('value')
+                    variant_data[namespace][feature].add(value)
+                data[VARIANTS_JSON_VARIANT_DATA_KEY][variant_config.vhash] = variant_data
+
+                def preprocess(data):
+                    """Preprocess the data to ensure it is JSON serializable."""
+                    if isinstance(data, (defaultdict, dict)):
```

Comment:
```suggestion
                    if isinstance(data, dict):
```

`defaultdict` is also a `dict`.